libs/base/Data/Colist.idr

```diff
@@ -0,0 +1,231 @@
+lappend xs = append (fromList xs)
+
+||| Append a `List` to a `Colist`.
+public export
+appendl : Colist a -> List a -> Colist a
+appendl xs = append xs . fromList
+
+||| Try to extract the first element from a `Colist`.
+public export
+head : Colist a -> Maybe a
+head []       = Nothing
+head (x :: _) = Just x
+
+||| Try to drop the first element from a `Colist`.
+||| This returns `Nothing` if the given `Colist` is
+||| empty.
+public export
+tail : Colist a -> Maybe (Colist a)
+tail []        = Nothing
+tail (_ :: xs) = Just xs
+
+||| Take up to `n` elements from a `Colist`.
+public export
+take : (n : Nat) -> Colist a -> List a
+take 0     _         = Nil
+take (S k) []        = Nil
+take (S k) (x :: xs) = x :: take k xs
+
+||| Take elements from a `Colist` up to and including the
+||| first element, for which `p` returns `True`.
+public export
+takeUntil : (p : a -> Bool) -> Colist a -> Colist a
+takeUntil _ []        = Nil
+takeUntil p (x :: xs) = if p x then [x] else x :: takeUntil p xs
+
+||| Take elements from a `Colist` up to (but not including) the
+||| first element, for which `p` returns `True`.
+public export
+takeBefore : (a -> Bool) -> Colist a -> Colist a
+takeBefore _ []        = Nil
+takeBefore p (x :: xs) = if p x then [] else x :: takeBefore p xs
+
+||| Take elements from a `Colist` while the given predicate
+||| returns `True`.
+public export
+takeWhile : (a -> Bool) -> Colist a -> Colist a
+takeWhile p = takeBefore (not . p)
+
+||| Extract all values wrapped in `Just` from the beginning
+||| of a `Colist`. This stops, once the first `Nothing` is encountered.
+public export
+takeWhileJust : Colist (Maybe a) -> Colist a
+takeWhileJust []              = []
+takeWhileJust (Nothing :: _)  = []
+takeWhileJust (Just x  :: xs) = x :: takeWhileJust xs
+
+||| Drop up to `n` elements from the beginning of the `Colist`.
+public export
+drop : (n : Nat) -> Colist a -> Colist a
+drop _ []            = Nil
+drop 0           xs  = xs
+drop (S k) (x :: xs) = drop k xs
+
+||| Try to extract the `n`-th element from a `Colist`.
+public export
+index : (n : Nat) -> Colist a -> Maybe a
+index _     []        = Nothing
+index 0     (x :: _)  = Just x
+index (S k) (_ :: xs) = index k xs
+
+||| Produce a `Colist` of left folds of prefixes of the given `Colist`.
+||| @ f the combining function
+||| @ acc the initial value
+||| @ xs the `Colist` to process
+public export
+scanl : (f : a -> b -> a) -> (acc : a) -> (xs : Colist b) -> Colist a
+scanl _ acc Nil       = [acc]
+scanl f acc (x :: xs) = acc :: scanl f (f acc x) xs
+
+--------------------------------------------------------------------------------
+--          Implementations
+--------------------------------------------------------------------------------
+
+public export
+Semigroup (Colist a) where
+  (<+>) = append
+
+public export
+Monoid (Colist a) where
+  neutral = Nil
+
+public export
+Functor Colist where
+  map f []        = []
+  map f (x :: xs) = f x :: map f xs
+
+public export
```

Comment:
Is it correct (or is it important) that [`Applicative` on `List`](https://github.com/idris-lang/Idris2/blob/master/libs/prelude/Prelude/Types.idr#L357) operates differently than this implementation on `Colist`?

on `List` it currently applies each function to each value and concats the results together.

Reply:
It looks like the current implementation mimics the `Data.Stream`'s implementation of `Applicative` and `Monad`. So, `Colist` is rather a `Stream` that can be finite, than a `List` that can be infinite.

Reply:
@mattpolzin: I think this is correct and important. Let's try and write a total implementation of `concatColist` in the spirit of `Data.List`

```
concatColist : Colist (Colist a) -> Colist a
concatColist Nil = Nil
concatColist ((x :: xs) :: xss) = x :: concatColist (xs :: xss)
concatColist (Nil :: xss) = concatColist xss
```

The last pattern match is problematic, since the recursive call is not guarded by `Inf`. In order to decide what the next element is, it might need to traverse the whole (possibly infinite) `Colist`. Indeed: `isNil . concatColist $ repeat Nil` will loop forever. Therefore, the module provides not `concatColist`, since for the time being I decided to only provide total functions.

But in order to implement `Applicative` and `Monad` in the `List` sense, we need something along the lines of `concatColist`.

The `Monad` instance for `Stream` behaves similarily: It treats a `Stream a` as a function `Nat -> a` (indeed: the two are isomorphic!) and the `Monad` instance of `Stream` behaves the same as the `Monad` instance for `Reader Nat`. This is, by the way, the same behavior like for `Vect n a` which is isomorphic to `Reader (Fin n) a` and whose `Monad` behaves the same as the one for `Reader (Fin n)`. Such data types are sometimes called _Naperian_ (see for instance [The Essense of the Iterator Pattern](https://www.cs.ox.ac.uk/jeremy.gibbons/publications/iterator.pdf)).

Now a `Colist` is not of a fixed size or shape, so it is not a Naperian data type. Still, it can be viewed as a function `index : Nat -> Maybe a` with the additional invariant that if `index n = Nothing` then `index (S n) = Nothing` and therefore its `Monad` and `Applicative` implementations behave like the ones for `ReaderT Nat Maybe`.

`Colist1` behaves the same with the only additional invariant that it always holds at least one element and therefore `index 0` always returns a `Just`.

However, all this writing made me realize I never verified the Applicative and Monad laws for these implementations. I'll give that a try in Idris and will report back here if I got stuck (or the laws are violated!).

Reply:
Thanks for the great reply! It makes intuitive sense to me now to implement it this way.

Reply:
I already got stuck proving the Functor laws, haha. I tried something along the following lines

```
I : a -> a
I a = a

functorId : (v : Colist a) -> (map I v = v)
functorId []       = Refl
functorId (x :: y) = cong2 (::) Refl ?foo
```

The problem her is, that `?foo` is cluttered with `Delay`s and `Force`s, and I have no idea how to get rid of those (its type is actually `Delay (map I Force y) = y`). Could someone please show me, if (and how) it is possible to properly deal with this? It would be great to have actual proofs of these laws in the module.

Reply:
Ok, I found the following [blog post](https://www.howtobuildsoftware.com/index.php/how-do/rS0/verification-idris-coinduction-proof-of-streams-functor-laws) describing a technique for proving equality of codata. In my case, this didn't work out in Idris2 (still problems with Delay and Force), but from this (and the isomorphisms mentioned above), I got the idea of using an element-wise equivalence: Two Colists can be considered equal, if `index` returns the same result for both of the for every natural number. Once again, I'll report back if this works out as expected.

Reply:
> The problem her is, that `?foo` is cluttered with `Delay`s and `Force`s, and I have no idea how to get rid of those (its type is actually `Delay (map I Force y) = y`). Could someone please show me, if (and how) it is possible to properly deal with this? It would be great to have actual proofs of these laws in the module.

You can match on a `Delay`, actually, to force it:
```idris
functorId : (xs : Colist a) -> map Prelude.id xs = xs
functorId [] = Refl
functorId (x::Delay xs) = cong (x::) $ functorId xs
```

Reply:
> I think this is correct and important

Sorry for repeating myself again, but I really think that the documentation comment on the `Colist` definition (that is currently "A possibly infinite list") is misleading and should be replaced with "A possibly finite Stream".

Reply:
@buzden Sorry, I didn't realize this was a comment regarding the data type's short description. I adjusted the descriptions accordingly.

Reply:
So, I found that the Monad laws for this implementation actually do not hold. Namely the law of associativity, which, expressed in terms of `join`, states that

```
join . join = join . map join
```

Minimal counter example:

```
Test : Colist (Colist (Colist Nat))
Test = [ [[1]], [[],[2,3]] ]

join Test = [[1],[2,3]]

join (join Test) = [1,3]

map join Test = [[1],[]]

join (map join Test) = [1]
```

I therefore removed the Monad implementations for Colist and Colist1.

Reply:
Yeah the right notion of equality on coinductive types is bisimilarity
(aka a coinductive notion of pointwise equality). Ohad has a student
incoming who will work on a prototype changing the notion of equality
in Idris to correspond to this (and change equality of functions to
pointwise equality too).

In the meantime, `(=#=)` in the linked post is the right thing to do.